raster/src/main/scala/geotrellis/raster/summary/GridVisitor.scala

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2019 Azavea
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package geotrellis.raster.summary
+
+import geotrellis.raster.Grid
+
+/**
+  * Visitor used to summarize a two-dimensional grid T that stores values of R
+  *
+  * The user should implement concrete subclasses that update the value of `result` as
+  * necessary on each call to `visit(raster: T, col: Int, row: Int)`.
+  *
+  * Be sure to handle the empty state. This could occur if no points in T are ever visited.
+  *
+  * GridVisitor is contravariant in T and covariant in R in the same fashion as Function1.
+  * This allows more generic concrete Visitor class implementations
+  * to satisfy type constraints for T at the call site, and the inverse is true for R. As a
+  * contrived example using GeoTrellis types, where Raster is a subtype of CellGrid:
+  *
+  * val DoubleVisitor: GridVisitor[CellGrid[Int], Double] = ???
+  * def usesRaster(visitor: GridVisitor[Raster[Tile]], Any]): Unit = ???
+  * usesRaster(DoubleVisitor)
+  *
+  * will compile. For an alternate explanation on covariance and contravariance that might help
+  * to clarify this, see:
+  * - https://stackoverflow.com/a/10604305
+  * - https://stackoverflow.com/a/38577878
+  */
+trait GridVisitor[-T <: Grid[Int], +R] {
```

Comment:
It looks like it could be relaxed to `-T <: Grid[_]` (everywhere in this PR we can be not that strict, at least without any obvious reason)

Reply:
Great point -- I believe we can. I'll make that change.

Reply:
Actually the `def visit(grid: T, col: Int, row: Int): Unit` implies that it has to be `Grid[Int]` and then the Rasterizer would only work on `(Int, Int)` grid.

So far semantically I've been using:
-  `Grid[Int]`  == Something that can be held in memory and addressed
- `Grid[Long]` == Something that is much larger, distributed and just describes the overall grid

So actually I'm not sure that it would be good idea to relax the constraint and I think you're going to run into a lot of problems as soon as you do it.

Reply:
@echeipesh Oops I didn't notice the `def visit(grid: T, col: Int, row: Int)` notion. Thanks. If it would be required, we'll have to add an extra hole into the `GridVisitor` declaration.

Reply:
I don't think thats worth it to add the type parameter. `Long`, `Long` addressable tiles would require a whole new addressing and paging strategy (since they're larger than memory) which doesn't exist in away that would be transparent to the `visitor`.

Reply:
Makes sense. So we leave as Grid[Int] for now.